File: medium/153-find-minimum-in-rotated-sorted-array.ts
// 

function findMin(nums: number[]): number {
  let left = 0, right = nums.length - 1;
  while (left < right) {
    let mid = left + Math.floor((right - left) / 2);
    if (nums[mid] > nums[right]) left = mid + 1;
    else right = mid;
  }
  return nums[left];
};

export default findMin;

File: python/153-find-minimum-in-rotated-sorted-array.py
from typing import List


def findMin(nums: List[int]) -> int:
    # l, r = 0, len(nums) - 1
    # while l < r:
    #     m = (r + l) // 2
    #     if nums[m] > nums[r]: l = m + 1
    #     else: r = m
    # return nums[l]

    low, high = 0, len(nums) - 1
    while low <= high:
        mid = (high + low) // 2
        if nums[mid] > nums[high]:
            low = mid + 1
        else:
            high = mid - 1
    return nums[low]


print(findMin([3, 4, 5, 1, 2]))
print(findMin([4, 5, 6, 7, 0, 1, 2]))
print(findMin([11, 13, 15, 17]))
